entrance_rando.py

```diff
@@ -0,0 +1,388 @@
+        # remove the placed targets from consideration
+        for entrance in removed_entrances:
+            entrance_lookup.remove(entrance)
+        # propagate new connections
+        er_state.collection_state.update_reachable_regions(world.player)
+        if on_connect:
+            on_connect(er_state, placed_exits)
+
+    def find_pairing(dead_end: bool, require_new_regions: bool) -> bool:
+        nonlocal perform_validity_check
+        placeable_exits = er_state.find_placeable_exits(perform_validity_check)
+        for source_exit in placeable_exits:
+            target_groups = target_group_lookup[source_exit.randomization_group]
+            for target_entrance in entrance_lookup.get_targets(target_groups, dead_end, preserve_group_order):
+                # requiring a new region is a proxy for enforcing new entrances are added, thus growing the search
+                # space. this is not quite a full fidelity conversion, but doesn't seem to cause issues enough
+                # to do more complex checks.
+                # the new region requirement can be ignored on a beaten minimal, islands are no issue there
+                region_requirement_satisfied = (not perform_validity_check or not require_new_regions
+                                                or target_entrance.connected_region not in er_state.placed_regions)
+                if region_requirement_satisfied and source_exit.can_connect_to(target_entrance, er_state):
+                    do_placement(source_exit, target_entrance)
+                    return True
+        else:
+            # no source exits had any valid target so this stage is deadlocked. retries may be implemented if early
+            # deadlocking is a frequent issue.
+            lookup = entrance_lookup.dead_ends if dead_end else entrance_lookup.others
+
+            # if we're in a stage where we're trying to get to new regions, we could also enter this
+            # branch in a success state (when all regions of the preferred type have been placed, but there are still
+            # additional unplaced entrances into those regions)
+            if require_new_regions:
+                if all(e.connected_region in er_state.placed_regions for e in lookup):
+                    return False
+
+            # if we're on minimal accessibility and can guarantee the game is beatable,
+            # we can prevent a failure by bypassing future validity checks. this check may be
+            # expensive; fortunately we only have to do it once
+            if perform_validity_check and world.options.accessibility == Accessibility.option_minimal \
+                    and world.multiworld.has_beaten_game(er_state.collection_state, world.player):
+                # ensure that we have enough locations to place our progression
+                accessible_location_count = 0
+                prog_item_count = len(er_state.collection_state.prog_items[world.player])
```

Comment:
`CollectionState.prog_items` is a **Dict** (specifically, Counter) per player.

This means that someone's `prog_items` look like this:

```
{
    "Sword": 3,
    "Shield": 1,
}
```

`len(prog_items[player])` would now return `2`, when there are `4` items collected.

This seems unintentional. It if isn't intentional, you'll want

```
sum(er_state.collection_state.prog_items[world.player].values()
```